python/tchannel/tornado/hyperbahn.py

```diff
@@ -22,18 +22,80 @@
 
 import json
 import logging
+import random
+from tornado.concurrent import Future
 
 import tornado.gen
 import tornado.ioloop
 
-from ..messages.error import ErrorCode
+from tchannel.tornado.timeout import timeout
 
-DEFAULT_TTL = 60  # seconds
+from ..errors import ProtocolError
+from ..errors import TimeoutError
+from .response import StatusCode
 
+DEFAULT_EXPO_BASE = 1.4  # try this first
+DEFAULT_MAX_DELAY = 10  # sec
+DEFAULT_MAX_ATTEMPT = 7  # pow(1.4, 8) > 10
+DEFAULT_DELAY = 60  # sec delay time for successful advertise
+DEFAULT_FIRST_ADVERTISE_TIME = 30  # sec
 
 log = logging.getLogger('tchannel')
 
 
+def _prepare_next_ad(attempt_counter):
+    if attempt_counter:
+        delay_time = random.uniform(
+            0, min(DEFAULT_MAX_DELAY, DEFAULT_EXPO_BASE ** attempt_counter)
+        )
+    else:
+        delay_time = DEFAULT_DELAY
+    attempt_counter = min(attempt_counter, DEFAULT_MAX_ATTEMPT)
+    return (attempt_counter, delay_time)
+
+
+@tornado.gen.coroutine
+def _advertise(tchannel, service, attempt_counter=0, recursion=False):
```

Comment:
`_advertise` shouldn't know anything about attempts, and we can also take this recursion bit out. The outer `advertise` method is going to handle the initial retry behavior, and then it's going to schedule a periodic callback.